File: src/web-ui/src/infrastructure/api/service-api/ReviewPlatformAPI.ts
import { api } from './ApiClient';
import { createTauriCommandError } from '../errors/TauriCommandError';
import { createLogger } from '@/shared/utils/logger';

const log = createLogger('ReviewPlatformAPI');

export type ReviewPlatformKind = 'github' | 'gitlab' | 'gitcode' | 'unknown';
export type ReviewAuthState = 'not_connected' | 'not_required' | 'connected' | 'expired' | 'error' | 'unsupported';
export type ReviewAuthSource = 'env' | 'stored' | 'none' | 'unsupported';
export type ReviewAuthChallengeState = 'missing' | 'invalid' | 'insufficient_scope';
export type ReviewItemState = 'open' | 'merged' | 'closed' | 'draft';
export type ReviewDecision = 'approved' | 'changes_requested' | 'commented' | 'pending';
export type ReviewFileStatus = 'added' | 'modified' | 'deleted' | 'renamed';
export type ReviewPlatformDetailSection = 'overview' | 'ci' | 'files' | 'commits' | 'reviews';

export interface ReviewPlatformAccount {
  id: string;
  platform: ReviewPlatformKind;
  label: string;
  username?: string | null;
  host: string;
  authState: ReviewAuthState;
  authSource: ReviewAuthSource;
  scopes: string[];
  message?: string | null;
}

export interface ReviewPlatformRepositoryRef {
  providerId: string;
  platform: ReviewPlatformKind;
  host: string;
  owner: string;
  name: string;
  projectPath: string;
  defaultBranch: string;
  workspacePath?: string | null;
  webUrl: string;
}

export interface ReviewPlatformAuthChallenge {
  platform: ReviewPlatformKind;
  host: string;
  remoteId: string;
  projectPath: string;
  state: ReviewAuthChallengeState;
  message: string;
  requiredScopes: string[];
}

export interface ReviewPlatformRemote {
  id: string;
  name: string;
  url: string;
  platform: ReviewPlatformKind;
  host: string;
  owner: string;
  repositoryName: string;
  projectPath: string;
  webUrl: string;
  supported: boolean;
  authState: ReviewAuthState;
  authSource: ReviewAuthSource;
  message?: string | null;
}

export interface ReviewChecks {
  total: number;
  passed: number;
  failed: number;
  pending: number;
}

export interface ReviewPlatformCiItem {
  id: string;
  name: string;
  status: string;
  conclusion?: string | null;
  detail?: string | null;
  stage?: string | null;
  webUrl?: string | null;
  log?: string | null;
  logTruncated: boolean;
  startedAt?: string | null;
  finishedAt?: string | null;
}

export interface ReviewPlatformPullRequest {
  id: string;
  number: number;
  title: string;
  state: ReviewItemState;
  author: string;
  sourceBranch: string;
  targetBranch: string;
  updatedAt: string;
  webUrl: string;
  additions: number;
  deletions: number;
  changedFiles: number;
  comments: number;
  reviewDecision: ReviewDecision;
  checks: ReviewChecks;
}

export interface ReviewPlatformFile {
  path: string;
  oldPath?: string | null;
  status: ReviewFileStatus;
  additions: number;
  deletions: number;
  patch?: string | null;
}

export interface ReviewPlatformCommit {
  hash: string;
  shortHash: string;
  title: string;
  author: string;
  committedAt: string;
}

export interface ReviewPlatformThread {
  id: string;
  providerThreadId?: string | null;
  providerCommentId?: string | null;
  kind: 'review' | 'comment';
  replyToProviderCommentId?: string | null;
  filePath?: string | null;
  line?: number | null;
  resolved: boolean;
  author: string;
  body: string;
  updatedAt: string;
}

export interface ReviewPlatformPullRequestDetail extends ReviewPlatformPullRequest {
  body: string;
  ci: ReviewPlatformCiItem[];
  files: ReviewPlatformFile[];
  commits: ReviewPlatformCommit[];
  threads: ReviewPlatformThread[];
}

export interface ReviewPlatformPullRequestDetailPage extends ReviewPlatformPullRequestDetail {
  section: ReviewPlatformDetailSection;
  pagination: ReviewPlatformPagination;
}

export interface ReviewPlatformCiLog {
  ciItemId: string;
  log?: string | null;
  truncated: boolean;
  message?: string | null;
}

export interface ReviewPlatformCapabilities {
  canCreateReview: boolean;
  canCreatePullRequest: boolean;
  canReplyToThread: boolean;
  canResolveThread: boolean;
  canApprove: boolean;
  canRevokeApproval: boolean;
  canRequestChanges: boolean;
  canMerge: boolean;
  supportsDraftReview: boolean;
}

export interface ReviewPlatformPagination {
  page: number;
  perPage: number;
  total?: number | null;
  hasNext: boolean;
}

export interface ReviewPlatformWorkspaceSnapshot {
  remotes: ReviewPlatformRemote[];
  selectedRemoteId?: string | null;
  accounts: ReviewPlatformAccount[];
  repository: ReviewPlatformRepositoryRef | null;
  pullRequests: ReviewPlatformPullRequest[];
  pagination: ReviewPlatformPagination;
  capabilities: ReviewPlatformCapabilities;
  message?: string | null;
  authChallenge?: ReviewPlatformAuthChallenge | null;
}

export interface ReviewPlatformWorkspaceSnapshotRequest {
  repositoryPath: string;
  remoteId?: string | null;
  page?: number;
  perPage?: number;
}

export interface ReviewPlatformPullRequestDetailRequest {
  repositoryPath: string;
  remoteId: string;
  pullRequestId: string;
}

export interface ReviewPlatformPullRequestDetailPageRequest extends ReviewPlatformPullRequestDetailRequest {
  section: ReviewPlatformDetailSection;
  page?: number;
  perPage?: number;
}

export interface ReviewPlatformPullRequestCiLogRequest extends ReviewPlatformPullRequestDetailRequest {
  ciItemId: string;
  ciItemName: string;
}

export interface ReviewPlatformUpdateAuthTokenRequest {
  platform: ReviewPlatformKind;
  host: string;
  token: string;
}

export interface ReviewPlatformClearAuthTokenRequest {
  platform: ReviewPlatformKind;
  host: string;
}

export class ReviewPlatformAPI {
  async getWorkspaceSnapshot(
    repositoryPath: string,
    remoteId?: string | null,
    page?: number,
    perPage?: number,
  ): Promise<ReviewPlatformWorkspaceSnapshot> {
    try {
      return await api.invoke('review_platform_get_workspace_snapshot', {
        request: { repositoryPath, remoteId, page, perPage },
      });
    } catch (error) {
      log.error('Failed to load review platform snapshot', { repositoryPath, remoteId, page, perPage, error });
      throw createTauriCommandError('review_platform_get_workspace_snapshot', error, {
        repositoryPath,
        remoteId,
        page,
        perPage,
      });
    }
  }

  async getPullRequestDetail(
    repositoryPath: string,
    remoteId: string,
    pullRequestId: string,
  ): Promise<ReviewPlatformPullRequestDetail> {
    try {
      return await api.invoke('review_platform_get_pull_request_detail', {
        request: { repositoryPath, remoteId, pullRequestId },
      });
    } catch (error) {
      log.error('Failed to load review platform pull request detail', {
        repositoryPath,
        remoteId,
        pullRequestId,
        error,
      });
      throw createTauriCommandError('review_platform_get_pull_request_detail', error, {
        repositoryPath,
        remoteId,
        pullRequestId,
      });
    }
  }

  async getPullRequestDetailPage(
    request: ReviewPlatformPullRequestDetailPageRequest,
  ): Promise<ReviewPlatformPullRequestDetailPage> {
    try {
      return await api.invoke('review_platform_get_pull_request_detail_page', {
        request,
      });
    } catch (error) {
      log.error('Failed to load review platform pull request detail page', {
        repositoryPath: request.repositoryPath,
        remoteId: request.remoteId,
        pullRequestId: request.pullRequestId,
        section: request.section,
        page: request.page,
        perPage: request.perPage,
        error,
      });
      throw createTauriCommandError('review_platform_get_pull_request_detail_page', error, request);
    }
  }

  async getPullRequestCiLog(
    request: ReviewPlatformPullRequestCiLogRequest,
  ): Promise<ReviewPlatformCiLog> {
    try {
      return await api.invoke('review_platform_get_pull_request_ci_log', {
        request,
      });
    } catch (error) {
      log.error('Failed to load review platform CI log', {
        repositoryPath: request.repositoryPath,
        remoteId: request.remoteId,
        pullRequestId: request.pullRequestId,
        ciItemId: request.ciItemId,
        error,
      });
      throw createTauriCommandError('review_platform_get_pull_request_ci_log', error, request);
    }
  }

  async updateAuthToken(request: ReviewPlatformUpdateAuthTokenRequest): Promise<void> {
    try {
      await api.invoke('review_platform_update_auth_token', { request });
    } catch (error) {
      log.error('Failed to update review platform auth token', {
        platform: request.platform,
        host: request.host,
        error,
      });
      throw createTauriCommandError('review_platform_update_auth_token', error, {
        platform: request.platform,
        host: request.host,
      });
    }
  }

  async clearAuthToken(request: ReviewPlatformClearAuthTokenRequest): Promise<void> {
    try {
      await api.invoke('review_platform_clear_auth_token', { request });
    } catch (error) {
      log.error('Failed to clear review platform auth token', {
        platform: request.platform,
        host: request.host,
        error,
      });
      throw createTauriCommandError('review_platform_clear_auth_token', error, {
        platform: request.platform,
        host: request.host,
      });
    }
  }
}

export const reviewPlatformAPI = new ReviewPlatformAPI();

File: src/apps/desktop/src/api/review_platform_api.rs
//! Review platform Tauri commands.

use crate::api::app_state::AppState;
use bitfun_core::service::review_platform::{
    ReviewPlatformCiLog, ReviewPlatformDetailSection, ReviewPlatformKind,
    ReviewPlatformPullRequestDetail, ReviewPlatformPullRequestDetailPage, ReviewPlatformService,
    ReviewPlatformWorkspaceSnapshot,
};
use log::error;
use serde::Deserialize;
use tauri::State;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformWorkspaceSnapshotRequest {
    pub repository_path: String,
    pub remote_id: Option<String>,
    pub page: Option<u32>,
    pub per_page: Option<u32>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformPullRequestDetailRequest {
    pub repository_path: String,
    pub remote_id: String,
    pub pull_request_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformPullRequestDetailPageRequest {
    pub repository_path: String,
    pub remote_id: String,
    pub pull_request_id: String,
    pub section: ReviewPlatformDetailSection,
    pub page: Option<u32>,
    pub per_page: Option<u32>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformPullRequestCiLogRequest {
    pub repository_path: String,
    pub remote_id: String,
    pub pull_request_id: String,
    pub ci_item_id: String,
    pub ci_item_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformUpdateAuthTokenRequest {
    pub platform: ReviewPlatformKind,
    pub host: String,
    pub token: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReviewPlatformClearAuthTokenRequest {
    pub platform: ReviewPlatformKind,
    pub host: String,
}

#[tauri::command]
pub async fn review_platform_get_workspace_snapshot(
    _state: State<'_, AppState>,
    request: ReviewPlatformWorkspaceSnapshotRequest,
) -> Result<ReviewPlatformWorkspaceSnapshot, String> {
    ReviewPlatformService::workspace_snapshot(
        &request.repository_path,
        request.remote_id.as_deref(),
        request.page,
        request.per_page,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform workspace snapshot: path={}, remote_id={:?}, error={}",
            request.repository_path, request.remote_id, error
        );
        format!(
            "Failed to get review platform workspace snapshot: {}",
            error
        )
    })
}

#[tauri::command]
pub async fn review_platform_get_pull_request_detail(
    _state: State<'_, AppState>,
    request: ReviewPlatformPullRequestDetailRequest,
) -> Result<ReviewPlatformPullRequestDetail, String> {
    ReviewPlatformService::pull_request_detail(
        &request.repository_path,
        &request.remote_id,
        &request.pull_request_id,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform pull request detail: path={}, remote_id={}, pull_request_id={}, error={}",
            request.repository_path,
            request.remote_id,
            request.pull_request_id,
            error
        );
        format!("Failed to get review platform pull request detail: {}", error)
    })
}

#[tauri::command]
pub async fn review_platform_get_pull_request_detail_page(
    _state: State<'_, AppState>,
    request: ReviewPlatformPullRequestDetailPageRequest,
) -> Result<ReviewPlatformPullRequestDetailPage, String> {
    ReviewPlatformService::pull_request_detail_page(
        &request.repository_path,
        &request.remote_id,
        &request.pull_request_id,
        request.section,
        request.page,
        request.per_page,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform pull request detail page: path={}, remote_id={}, pull_request_id={}, section={:?}, page={:?}, per_page={:?}, error={}",
            request.repository_path,
            request.remote_id,
            request.pull_request_id,
            request.section,
            request.page,
            request.per_page,
            error
        );
        format!(
            "Failed to get review platform pull request detail page: {}",
            error
        )
    })
}

#[tauri::command]
pub async fn review_platform_get_pull_request_ci_log(
    _state: State<'_, AppState>,
    request: ReviewPlatformPullRequestCiLogRequest,
) -> Result<ReviewPlatformCiLog, String> {
    ReviewPlatformService::pull_request_ci_log(
        &request.repository_path,
        &request.remote_id,
        &request.pull_request_id,
        &request.ci_item_id,
        &request.ci_item_name,
    )
    .await
    .map_err(|error| {
        error!(
            "Failed to get review platform CI log: path={}, remote_id={}, pull_request_id={}, ci_item_id={}, error={}",
            request.repository_path,
            request.remote_id,
            request.pull_request_id,
            request.ci_item_id,
            error
        );
        format!("Failed to get review platform CI log: {}", error)
    })
}

#[tauri::command]
pub async fn review_platform_update_auth_token(
    _state: State<'_, AppState>,
    request: ReviewPlatformUpdateAuthTokenRequest,
) -> Result<(), String> {
    ReviewPlatformService::update_auth_token(request.platform, &request.host, &request.token)
        .await
        .map_err(|error| {
            error!(
                "Failed to update review platform auth token: platform={:?}, host={}, error={}",
                request.platform, request.host, error
            );
            format!("Failed to update review platform auth token: {}", error)
        })
}

#[tauri::command]
pub async fn review_platform_clear_auth_token(
    _state: State<'_, AppState>,
    request: ReviewPlatformClearAuthTokenRequest,
) -> Result<(), String> {
    ReviewPlatformService::clear_auth_token(request.platform, &request.host)
        .await
        .map_err(|error| {
            error!(
                "Failed to clear review platform auth token: platform={:?}, host={}, error={}",
                request.platform, request.host, error
            );
            format!("Failed to clear review platform auth token: {}", error)
        })
}
